src/closure.rs

```diff
@@ -0,0 +1,163 @@
+// TODO: support marshaller.
+
+use std::mem;
+use std::ptr;
+
+use libc::{c_uint, c_void};
+
+use gobject_ffi;
+use translate::{FromGlibPtrFull, FromGlibPtrNone, Stash, StashMut, ToGlibPtr, ToGlibPtrMut, Uninitialized};
+use Value;
+use ToValue;
+
+pub struct Closure {
+    closure: *mut gobject_ffi::GClosure,
+}
+
+impl Closure {
+    pub fn new<F: Fn(&[Value]) -> Option<Value> + 'static>(callback: F) -> Self {
+        unsafe extern "C" fn marshal<F: Fn(&[Value]) -> Option<Value> + 'static>(_closure: *mut gobject_ffi::GClosure,
+            return_value: *mut gobject_ffi::GValue, n_param_values: c_uint, mut param_values: *mut gobject_ffi::GValue,
+            _invocation_hint: *mut c_void, marshal_data: *mut c_void)
+        {
+            let mut values = vec![];
+            for _ in 0 .. n_param_values {
+                let value: Value = FromGlibPtrNone::from_glib_none(param_values);
+                values.push(value);
+                param_values = param_values.offset(1);
+            }
+            let callback: &Box<F> = mem::transmute(marshal_data as *mut _);
```

Comment:
You can also use Box::from_raw() here and just mem::forget() it before returning.

You should also create a glib::CallbackGuard() here to catch panics and exit on panics